README.md
## Eero Device Tracker for Home Assistant

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![hacs_badge](https://img.shields.io/badge/HACS-Default-orange.svg)](https://github.com/custom-components/hacs)

Adds device tracking support for [Eero Mesh WiFi routers](https://eero.com/) to [Home Assistant](https://www.home-assistant.io/).

This project is based on [@343max's eero-client project](https://github.com/343max/eero-client) -- many thanks to him or this project probably wouldn't exist. Thanks to [@rsnodgrass](https://github.com/rsnodgrass) for the structural improvements. This code has no warranties, and please submit any pull requests you might have to improve upon it.

## Setup Process

1. Install using [HACS](https://github.com/hacs/integration) or manually copy the files
2. Run the eero_tracker_instantiate.py script to setup credentials (creates an eero.session in /config)
3. Add eero_tracker to your configuration.yaml
4. Restart Home Assistant

Now let's get on with installin' this thing!

## Step 1: Installation

### Option A: Installation with Home Assistant Community Store (HACS)

For easy updates whenever a new version is released, use the [Home Assistant Community Store (HACS)](https://github.com/hacs/integration) and add the following Integration in the Settings tab:

```
jrlucier/eero_tracker
```

Then go to the Integration tab, search for "Eero Tracker", and click install.

### Option B: Manual Installation

Copy the scripts!

SSH into your device. Download the zip file from the releases section and uncompress it in your configuration directory (e.g. `~/.homeassistant` or `/config`):

```
cd ~/.homeassistant/
wget https://github.com/jrlucier/eero_tracker/releases/download/1.0.10/eero_tracker-1.0.10.zip
unzip eero_tracker-1.0.10.zip
```

## Step 2: Generate Credentials for Connecting to Your Eeros

#### Why do I need to run a script in my config directory?

Eero doesn't have a traditional user/password login setup, so we need to use your phone's SMS or your email address to create an authenticated session token (`eero.session`). To do that, you'll need to SSH into your device running Home Assistant to run an interactive script that allows that to occur.

#### Note for Hass.io users

If you're running [Hass.io](https://www.home-assistant.io/hassio/), be aware that the official SSH server will not allow you to run python files (so I'm told, and which we require), so use the Secure Shell community add-on. Your configuration directory will be stored under `/config` instead of `~/.homeassistant/`.

An alternative method would be to run `eero_tracker_instantiate.py` on another machine, and then manually copy over the `eero.session` file to the config directory on the Home Assistant host.

#### Running the Script

The scripts in this project rely on the "requests" package, so install it if it's not already installed:

```
python3 -m pip install requests
```

We need to get an authenticated session created with Eero's servers. So to do that, you'll need to go to your configuration directory (eg: `cd ~/.homeassistant`), and run the `eero_tracker_instantiate.py` file:

```
python3 eero_tracker_instantiate.py
```

This will prompt you for your phone number (no dashes), and then it will send you an SMS text with a code you will need to put in. You may also use an email address instead. Once done, it will create an `eero.session` file in your configuration directory.  Subsequent calls to this python file will dump the list of connected wireless devices, their mac addresses, and hostnames for easier reference.  You technically shouldn't need `eero_tracker_instantiate.py` after the creation of the `eero.session` file, but I keep it around for quick mac address referencing.

#### Manual Installation Permissions

If you aren't running [Hass.io](https://www.home-assistant.io/hassio/) (whose default SSH user is root), and have Home Assistant configured differently, then check the permissions on the files. `chown` the files to the same permissions as your other HA configuration files (`ls -al` to check yours in your configuration directory). Mine are owned by `homeassistant:nogroup`:

```bash
sudo chown homeassistant:nogroup eero.session
sudo chown homeassistant:nogroup eero_tracker_instantiate.py 
sudo chown -R homeassistant:nogroup custom_components/
```

## Step 3: Add Tracker to Home Assistant's Configuration

Now that that installation and authentication are done, all that is left is to add the [device_tracker](https://www.home-assistant.io/integrations/device_tracker/) to your `configuration.yaml`.

The minimum required configuration:

```yaml
device_tracker:
  - platform: eero_tracker
```

Example including optional configuration:

```yaml
device_tracker:
  - platform: eero_tracker

    # optional eero_tracker settings:
    only_macs: "11:22:33:44:55:66,22:22:22:22:22:22"
    only_networks:
      - 30120

    # standard device_tracker options
    consider_home: 300     # default: 180
    interval_seconds: 30   # default: 25
    only_wireless: True    # default: True
```

#### Config Keys

| Key                | Default | Description |
|--------------------|---------|-------------|
| `only_macs`        | none    | comma separated list of MAC addresses that reduces the devices monitored to a smaller set. |
| `interval_seconds` | 180     | **must** be 25 seconds or greater to avoid DDoS of eero's servers. |
| `only_networks`    | none    | YAML list of network identifiers to search for devices (only useful if you have multiple eero locations under a single eero email address, for instance at work or a second home). Turn on HA debug logging to determine the network ids for your eeros |
| `only_wireless`    | True    | only track wireless devices if set to true (normally hardwired devices are not useful for tracking)

For additional device tracker configuration options, see the [HA device_tracker docs](https://www.home-assistant.io/integrations/device_tracker/).

## Step 4: Restart and Test

You should see wireless devices populate using each device's nicknames, where possible, as the device name.

**NOTE: This does not populate any devices that are not wirelessly connected to your eero.**

## Support

If you are experiencing any issues, first check the [community support discussion thread](https://community.home-assistant.io/t/eero-support/21153) to see if anyone else has solved your issue previously. You can also discuss the issue you are having there. If you feel it is a bug, please [create an github Issue with the details](https://github.com/jrlucier/eero_tracker/issues).

#### Not Yet Implemented

The following features are not yet implemented (no plans currently for adding). If you are interested in contributing code, please submit a patch.

- support for family profiles (pause/unpause switch) and assigning to Home Assistant "person" entities
- support for rebooting the eero network
- eero connection status and most recent upload/download speed test results
- config_flow to allow all configuration through the Home Assistant UI (including eero.session setup)

## See Also

* [Eero Tracker community discussion forum](https://community.home-assistant.io/t/eero-support/21153)
* [Eero Mesh WiFi routers](https://eero.com/) (official site)
* [Eero Python Client (343max/eero-client)](https://github.com/343max/eero-client)
* [Eero API endpoint examples](https://github.com/yepher/eeroMonitor)


CHANGELOG.md
## 1.0.10 (2021-05-07)
New config flow thanks to Jack Stevenson (github: cogwirrel)
Fixed HA versioning error thanks to Ron Heft (github: ronaldheft)

## 1.0.6 (2020-03-01)
Changes to device_tracker since there's been some deprecation and changing of config's usage.

## 1.0.3 (2019-12-26)
Updates with support for installing using HACS

## 1.0.2 (2019-02-23)
API changes

## 1.0.1 (2018-05-18)
Initial release of Eero tracker


eero_tracker_instantiate.py
import re
import json
import requests
from argparse import ArgumentParser
from abc import abstractproperty

__version__ = "0.0.1"
__all__ = ['ClientException', 'Eero', 'SessionStorage', '__version__']


class Eero(object):
    def __init__(self, arg_session):
        # type(SessionStorage) -> ()
        self.session = arg_session
        self.client = Client()

    @property
    def _cookie_dict(self):
        if self.needs_login():
            return dict()
        else:
            return dict(s=self.session.cookie)

    def needs_login(self):
        return self.session.cookie is None

    def login(self, identifier):
        # type(string) -> string
        json = dict(login=identifier)
        data = self.client.post('login', json=json)
        return data['user_token']

    def login_verify(self, verification_code, user_token):
        json = dict(code=verification_code)
        response = self.client.post('login/verify', json=json,
                                    cookies=dict(s=user_token))
        self.session.cookie = user_token
        return response

    def refreshed(self, func):
        try:
            return func()
        except ClientException as exception:
            if exception.status == 401 and exception.error_message == 'error.session.refresh':
                self.login_refresh()
                return func()
            else:
                raise

    def login_refresh(self):
        response = self.client.post('login/refresh', cookies=self._cookie_dict)
        self.session.cookie = response['user_token']

    def account(self):
        return self.refreshed(lambda: self.client.get(
            'account',
            cookies=self._cookie_dict))

    @staticmethod
    def id_from_url(id_or_url):
        match = re.search('^[0-9]+$', id_or_url)
        if match:
            return match.group(0)
        match = re.search(r'\/([0-9]+)$', id_or_url)
        if match:
            return match.group(1)

    def devices(self, network_id):
        return self.refreshed(lambda: self.client.get(
            'networks/{}/devices'.format(
                self.id_from_url(network_id)), cookies=self._cookie_dict))


class SessionStorage(object):
    @abstractproperty
    def cookie(self):
        pass


class ClientException(Exception):
    def __init__(self, status, error_message):
        super(ClientException, self).__init__()
        self.status = status
        self.error_message = error_message


class Client(object):
    API_ENDPOINT = 'https://api-user.e2ro.com/2.2/{}'

    @staticmethod
    def _parse_response(response):
        data = json.loads(response.text)
        if data['meta']['code'] != 200 and data['meta']['code'] != 201:
            raise ClientException(data['meta']['code'],
                                  data['meta'].get('error', ""))
        return data.get('data', "")

    def post(self, action, **kwargs):
        response = requests.post(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)

    def get(self, action, **kwargs):
        response = requests.get(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)


class CookieStore(SessionStorage):
    def __init__(self, cookie_file):
        from os import path
        self.cookie_file = path.abspath(cookie_file)

        try:
            with open(self.cookie_file, 'r') as f:
                self.__cookie = f.read()
        except IOError:
            self.__cookie = None

    @property
    def cookie(self):
        return self.__cookie

    @cookie.setter
    def cookie(self, cookie):
        self.__cookie = cookie
        with open(self.cookie_file, 'w+') as f:
            f.write(self.__cookie)


session = CookieStore('eero.session')
eero = Eero(session)

if __name__ == '__main__':
    if eero.needs_login():

        try:
            input = raw_input
        except NameError:
            pass

        parser = ArgumentParser()
        parser.add_argument("-l", help="Your eero login (email address or SMS phone number)")
        args = parser.parse_args()
        if args.l:
            phone_number = args.l
        else:
            phone_number = input('Your eero login (email address or SMS phone number): ')
        user_token = eero.login(phone_number)
        verification_code = input('Verification key from email or SMS: ')
        eero.login_verify(verification_code, user_token)
        print('Login successful. eero.session created, you can now use the device_tracker.')
    else:
        print('eero.session already created dumping all devices')
        account = eero.account()
        for network in account['networks']['data']:
            devices = eero.devices(network['url'])

            json_obj = json.loads(json.dumps(devices, indent=4))
            for device in json_obj:
                if device['wireless'] and device['connected']:
                    print("{}, {}, {}".format(device['nickname'], device['hostname'], device['mac']))


custom_components/eero_tracker/eero.py
import re
import json
import requests
from argparse import ArgumentParser
from abc import abstractproperty


class Eero(object):
    def __init__(self, arg_session):
        # type(SessionStorage) -> ()
        self.session = arg_session
        self.client = Client()

    @property
    def _cookie_dict(self):
        if self.needs_login():
            return dict()
        else:
            return dict(s=self.session.cookie)

    def needs_login(self):
        return self.session.cookie is None

    def login(self, identifier):
        # type(string) -> string
        json = dict(login=identifier)
        data = self.client.post('login', json=json)
        return data['user_token']

    def login_verify(self, verification_code, user_token):
        json = dict(code=verification_code)
        response = self.client.post('login/verify', json=json,
                                    cookies=dict(s=user_token))
        self.session.cookie = user_token
        return response

    def refreshed(self, func):
        try:
            return func()
        except ClientException as exception:
            if exception.status == 401 and exception.error_message == 'error.session.refresh':
                self.login_refresh()
                return func()
            else:
                raise

    def login_refresh(self):
        response = self.client.post('login/refresh', cookies=self._cookie_dict)
        self.session.cookie = response['user_token']

    def account(self):
        return self.refreshed(lambda: self.client.get(
            'account',
            cookies=self._cookie_dict))

    @staticmethod
    def id_from_url(id_or_url):
        match = re.search('^[0-9]+$', id_or_url)
        if match:
            return match.group(0)
        match = re.search(r'\/([0-9]+)$', id_or_url)
        if match:
            return match.group(1)

    def devices(self, network_id):
        return self.refreshed(lambda: self.client.get(
            'networks/{}/devices'.format(
                self.id_from_url(network_id)), cookies=self._cookie_dict))


class SessionStorage(object):
    @abstractproperty
    def cookie(self):
        pass


class ClientException(Exception):
    def __init__(self, status, error_message):
        super(ClientException, self).__init__()
        self.status = status
        self.error_message = error_message


class Client(object):
    API_ENDPOINT = 'https://api-user.e2ro.com/2.2/{}'

    @staticmethod
    def _parse_response(response):
        data = json.loads(response.text)
        if data['meta']['code'] != 200 and data['meta']['code'] != 201:
            raise ClientException(data['meta']['code'],
                                  data['meta'].get('error', ""))
        return data.get('data', "")

    def post(self, action, **kwargs):
        response = requests.post(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)

    def get(self, action, **kwargs):
        response = requests.get(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)


class CookieStore(SessionStorage):
    def __init__(self, cookie_file):
        from os import path
        self.cookie_file = path.abspath(cookie_file)

        try:
            with open(self.cookie_file, 'r') as f:
                self.__cookie = f.read()
        except IOError:
            self.__cookie = None

    @property
    def cookie(self):
        return self.__cookie

    @cookie.setter
    def cookie(self, cookie):
        self.__cookie = cookie
        with open(self.cookie_file, 'w+') as f:
            f.write(self.__cookie)


custom_components/eero_tracker/config_flow.py
import logging
import voluptuous as vol

from collections import OrderedDict
from homeassistant import config_entries

from .const import (
    DOMAIN,
    EERO_SESSION_COOKIE_FILE,
)

from .eero import (
    CookieStore,
    Eero,
)

_LOGGER = logging.getLogger(__name__)

@config_entries.HANDLERS.register(DOMAIN)
class EeroFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):

    def __init__(self):
        self.session = CookieStore(EERO_SESSION_COOKIE_FILE)
        self.eero = Eero(self.session)
        self.token = None
        self.username = None

        self.user_step_schema = vol.Schema({ vol.Required("username"): str })
        self.verify_step_schema = vol.Schema({ vol.Required("verification_code"): str })

    async def async_step_user(self, user_input=None):
        _LOGGER.debug("Handle user step with input", user_input)
        if user_input is not None:
            self.username = user_input["username"]
            _LOGGER.debug("Starting login for user", self.username)
            await self.async_set_unique_id(self.username)
            self.token = self.eero.login(self.username)
            _LOGGER.debug("Retrieved token", self.token)
            return self.async_show_form(
                step_id="verify", data_schema=self.verify_step_schema
            )

        _LOGGER.debug("Starting login flow")
        return self.async_show_form(
            step_id="user", data_schema=self.user_step_schema
        )

    async def async_step_verify(self, user_input=None):
        _LOGGER.debug("Verifiying user")
        self.eero.login_verify(user_input["verification_code"], self.token)
        _LOGGER.debug("Verification successful")
        return self.async_create_entry(
            title=f"Eero - {self.username}",
            data={"username": self.username}
        )


custom_components/eero_tracker/device_tracker.py
"""
Eero WiFi router device_tracker for Home Assistant

For instructions and examples, see https://github.com/jrlucier/eero_tracker
"""
import logging
import voluptuous as vol
import datetime
import time
import re
import json
import requests
import homeassistant.helpers.config_validation as cv
from homeassistant.components.device_tracker.legacy import DeviceScanner
from homeassistant.components.device_tracker import PLATFORM_SCHEMA
from homeassistant.components.device_tracker.const import (
           DOMAIN, CONF_SCAN_INTERVAL)

_LOGGER = logging.getLogger(__name__)

CONF_ONLY_MACS_KEY = 'only_macs'
CONF_ONLY_NETWORKS = 'only_networks'
CONF_ONLY_WIRELESS = 'only_wireless'
CONF_SESSION_FILE_NAME = 'session_file_name'

MINIMUM_SCAN_INTERVAL = 25

CACHE_EXPIRY=3600 # cache accounts for an hour

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    vol.Optional(CONF_ONLY_MACS_KEY, default=''): cv.string,
    vol.Optional(CONF_ONLY_NETWORKS, default=[]): vol.All(cv.ensure_list, [cv.positive_int]),
    vol.Optional(CONF_ONLY_WIRELESS, default=True): cv.boolean, 
    vol.Optional(CONF_SESSION_FILE_NAME, default='eero.session'): cv.string
})

def get_scanner(hass, config):
    """Validate the configuration and return EeroDeviceScanner."""

    _LOGGER.debug(f"Initializing eero_tracker (domain {DOMAIN})")
    return EeroDeviceScanner(hass, config[DOMAIN])

class EeroException(Exception):
    """A propagating error for Eero"""

    def __init__(self, status, error_message):
        super(EeroException, self).__init__()
        self.status = status
        self.error_message = error_message


class EeroDeviceScanner(DeviceScanner):
    """This class queries a Eero-based router for present devices."""

    API_ENDPOINT = 'https://api-user.e2ro.com/2.2/{}'

    def __init__(self, hass, config):
        """Initialize the scanner."""
        self.__session_file = hass.config.path(config[CONF_SESSION_FILE_NAME])
        self.__session = None
        
        # configure any filters (macs or networks)
        self.__only_macs = set([x.strip().lower() for x in config[CONF_ONLY_MACS_KEY].split(',') if x != ''])
        if len(self.__only_macs) > 0:
            _LOGGER.info(f"Including only MAC addresses: {self.__only_macs}")

        self.__only_networks = set(config[CONF_ONLY_NETWORKS])
        if len(self.__only_networks) > 0:
            _LOGGER.info(f"Including only networks: {self.__only_networks}")

        self.__only_wireless = config[CONF_ONLY_WIRELESS]
        _LOGGER.info(f"Tracking only wireless devices = {self.__only_wireless}")

        self.__last_results = []
        self.__account = None
        self.__account_update_timestamp = None
        self.__mac_to_nickname = {}
        self.__mac_to_attrs = {}           

        minimum_interval = datetime.timedelta(seconds=MINIMUM_SCAN_INTERVAL)
        self.__scan_interval = config.get(CONF_SCAN_INTERVAL, minimum_interval)

        # Prevent users from specifying an interval faster than 25 seconds
        if self.__scan_interval < minimum_interval:
            _LOGGER.warning(
                f"Scan interval {self.__scan_interval} MUST be >= {MINIMUM_SCAN_INTERVAL} seconds to prevent DDoS on eero's servers; limiting to {minimum_interval}.")
            self.__scan_interval = minimum_interval
        else:
            _LOGGER.debug(f"Scan interval = {self.__scan_interval}")

        # Grab the session key from the file
        try:
            _LOGGER.debug(f"Loading eero session key from '{self.__session_file}'")
            with open(self.__session_file, 'r') as f:
                self.__session = f.read().replace('\n', '')
        except IOError:
            _LOGGER.error(f"Could not find the eero.session file '{self.__session_file}'")
            self.__session = None

    def scan_devices(self):
        """Required for the API, handles returning results"""
        # Return empty array if the session was never started.
        if self.__session is None:
            return []

        self._update_info()
        return self.__last_results

    def get_device_name(self, mac):
        """Required for the API. None to indicate we don't know the devices true name"""
        return self.__mac_to_nickname.get(mac)

    def get_extra_attributes(self, mac):
        """Get the extra attributes of a device."""
        return self.__mac_to_attrs.get(mac)

    def _update_info(self):
        """Retrieve the latest information from Eero for returning to HA."""
        # Cache the accounts for an hour. These rarely change and this reduces the
        # lookup requests to only 1 every update. This cache is reset on Home Assistant
        # restarts, so in an emergency a user can always restart Home Assistant to force update.
        if self.__account_update_timestamp is None or (time.time() - self.__account_update_timestamp) >= CACHE_EXPIRY:
            _LOGGER.debug(f"Updating eero account information cache (expires every {CACHE_EXPIRY} seconds)")
            self.__account = self._account()
            self.__account_update_timestamp = time.time()

        self.__mac_to_nickname = {}
        self.__mac_to_attrs = {}
        self.__last_results = []
        
        for network in self.__account['networks']['data']:
            match = re.search('/networks/(\d+)', network['url'])
            network_id = int(match.group(1))
            
            # if specific networks should be filtered, skip any not in the filter
            if len(self.__only_networks) > 0 and network_id not in self.__only_networks:
                _LOGGER.debug(f"Ignoring network {network_id} devices not in only_networks: {self.__only_networks}")
                continue

            # load all devices for this network, but only track connected wireless devices
            devices = self._devices(network['url'])
            json_obj = json.loads(json.dumps(devices, indent=4))
            self._update_tracked_devices(network_id, json_obj)

        return

    def _update_tracked_devices(self, network_id, devices_json_obj):
        for device in devices_json_obj:
            # skip devices that are not connected
            if not device['connected']:
                continue

            # if only wireless devices are tracked, then skip if not wireless
            if self.__only_wireless and not device['wireless']:
                continue

            # if mac addressess are whitelisted with only_macs, skip if not on the list
            mac = device['mac']
            if len(self.__only_macs) > 0 and mac not in self.__only_macs:
                continue

            # create mapping of mac addresses to nicknames for lookup by device_name (if a nickname is assigned)
            nickname = device['nickname']

            # default nickname to host name if missing
            if not nickname or nickname == 'None':
                nickname = device['hostname']

            if nickname:
                self.__mac_to_nickname[mac] = nickname

            attrs = device["source"]
            if attrs:
                self.__mac_to_attrs[mac] = device["source"]

            _LOGGER.debug(f"Network {network_id} device found: nickname={nickname}; host={device['hostname']}; mac={mac}")
            self.__last_results.append(mac)

    @property
    def _cookie_dict(self):
        """Creates a session cookie"""
        return dict(s=self.__session)

    def _refreshed(self, func):
        """Handles if we need to refresh the logged in session or not"""
        try:
            return func()
        except EeroException as exception:
            if exception.status == 401 and exception.error_message == 'error.session.refresh':
                self._login_refresh()
                return func()
            else:
                _LOGGER.error(f"Eero connection failure: {exception.error_message}")

    def _login_refresh(self):
        """Refresh the Eero session"""
        response = self._post_req('login/refresh', cookies=self._cookie_dict)
        new_session = response.get('user_token')
        if not new_session:
            _LOGGER.error(f"Failed updating eero session key! {response}")
            return

        _LOGGER.debug(f"Updating {self.__session_file} with new session key")
        try:
            # update in-memory session first, in case there is any failure in writing to the
            # session file, at least this tracker will continue working until next HA restart
            self.__session = new_session

            # TODO: ideally write to a temp file, and if successful, then move to overwrite
            # the existing session file
            with open(self.__session_file, 'w+') as f:
                f.write(new_session)
        except IOError:
            _LOGGER.error(f"Could not update eero session key in {self.__session_file}")

    def _account(self):
        return self._refreshed(lambda: self._get_req('account', cookies=self._cookie_dict))

    @staticmethod
    def _id_from_url(id_or_url):
        """Handles grabbing the Eero ID from the URL"""
        match = re.search('^[0-9]+$', id_or_url)
        if match:
            return match.group(0)
        match = re.search(r'\/([0-9]+)$', id_or_url)
        if match:
            return match.group(1)

    def _devices(self, network_id):
        """Gets the list of devices from Eero"""
        return self._refreshed(lambda: self._get_req('networks/{}/devices'.format(self._id_from_url(network_id)),
                                                     cookies=self._cookie_dict))

    @staticmethod
    def _parse_response(response):
        """Basic response handler"""
        data = json.loads(response.text)
        response_code = data['meta']['code']
        if response_code not in [ 200, 201 ]:
            raise EeroException(response_code, data['meta'].get('error', ""))
        return data.get('data', "")

    def _post_req(self, action, **kwargs):
        """POST a request"""
        response = requests.post(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)

    def _get_req(self, action, **kwargs):
        """GET a request"""
        response = requests.get(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)


custom_components/device_tracker/eero_tracker.py
"""
Support for Eero routers.

Place this eero_tracker.py file in: ~/.homeassistant/custom_components/device_tracker/

Given a session which can be located via the original eero.py file.  Here's an example configuration.yaml for this:
device_tracker:
  - platform: eero_tracker
    consider_home: 300
    interval_seconds: 60
    only_macs: "11:22:33:44:55:66, 22:22:22:22:22:22"

"""
import logging
import voluptuous as vol
import datetime
import time
import re
import json
import requests
import homeassistant.helpers.config_validation as cv
from homeassistant.components.device_tracker import (
    DOMAIN, PLATFORM_SCHEMA, CONF_SCAN_INTERVAL, DeviceScanner)

REQUIREMENTS = ['requests==2.13.0']

_LOGGER = logging.getLogger(__name__)

CONF_ONLY_MACS_KEY = 'only_macs'
CONF_SESSION_FILE_NAME = 'session_file_name'

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    vol.Optional(CONF_ONLY_MACS_KEY, default=''): cv.string,
    vol.Optional(CONF_SESSION_FILE_NAME, default='eero.session'): cv.string
})


def get_scanner(hass, config):
    """Validate the configuration and return EeroDeviceScanner."""

    _LOGGER.debug('Eero init')
    return EeroDeviceScanner(hass, config[DOMAIN])


class EeroException(Exception):
    """A propagating error for Eero"""

    def __init__(self, status, error_message):
        super(EeroException, self).__init__()
        self.status = status
        self.error_message = error_message


class EeroDeviceScanner(DeviceScanner):
    """This class queries a Eero-based router for present devices."""

    API_ENDPOINT = 'https://api-user.e2ro.com/2.2/{}'

    def __init__(self, hass, config):
        """Initialize the scanner."""
        self.__session_file = hass.config.path(config[CONF_SESSION_FILE_NAME])
        self.__session = None
        self.__only_macs = set([x.strip().lower() for x in config[CONF_ONLY_MACS_KEY].split(',') if x != ''])
        self.__scan_interval = config[CONF_SCAN_INTERVAL]
        self.__last_results = []
        self.__account = None
        self.__account_update_timestamp = None
        self.__mac_to_nickname = {}

        # Prevent users from specifying an interval faster than 25.  This will block the tracker
        # from returning results by bailing out early
        if self.__scan_interval < datetime.timedelta(seconds=25):
            _LOGGER.error(
                'Disabled. Scan interval is too fast!  Must be 25 or greater to prevent DDOSing Eeros servers.')
            return

        # Grab the session from the file
        try:
            with open(self.__session_file, 'r') as f:
                self.__session = f.read().replace('\n', '')
        except IOError:
            _LOGGER.error('Could not find the eero.session file at: {}'.format(self.__session_file))
            self.__session = None

    def scan_devices(self):
        """Required for the API, handles returning results"""
        # Return empty array if the session was never started...usually by too fast a speed specified.
        if self.__session is None:
            return []

        self._update_info()
        return self.__last_results

    def get_device_name(self, mac):
        """Required for the API. None to indicate we don't know the devices true name"""
        return self.__mac_to_nickname.get(mac)

    def _update_info(self):
        """Retrieve the latest information from Eero for returning to HA."""
        # Cache the accounts for an hour...these should rarely change and reduces the lookup requests
        # we use to only 1 every update.
        if self.__account is None or self.__account_update_timestamp is None or \
                        (time.time() - self.__account_update_timestamp) > 3600:
            _LOGGER.debug("Updating eero account information cache...")
            self.__account = self._account()
            self.__account_update_timestamp = time.time()

        self.__mac_to_nickname = {}
        self.__last_results = []
        for network in self.__account['networks']['data']:
            devices = self._devices(network['url'])

            json_obj = json.loads(json.dumps(devices, indent=4))
            for device in json_obj:
                if device['wireless'] and device['connected']:

                    if len(self.__only_macs) > 0 and device['mac'] not in self.__only_macs:
                        continue

                    _LOGGER.debug(
                        "Device found: {}, {}, {}".format(device['nickname'], device['hostname'], device['mac']))

                    # Create a mapping of macs to nicknames for lookup by device_name, if a nickname is assigned
                    if device['nickname']:
                        self.__mac_to_nickname[device['mac']] = device['nickname']

                    # Append a result
                    self.__last_results.append(device['mac'])

        return

    @property
    def _cookie_dict(self):
        """Creates a session cookie"""
        return dict(s=self.__session)

    def _refreshed(self, func):
        """Handles if we need to refresh the logged in session or not"""
        try:
            return func()
        except EeroException as exception:
            if exception.status == 401 and exception.error_message == 'error.session.refresh':
                self._login_refresh()
                return func()
            else:
                _LOGGER.error('Eero connection failure: {}'.format(exception.error_message))

    def _login_refresh(self):
        """Refresh the Eero session"""
        response = self._post_req('login/refresh', cookies=self._cookie_dict)
        self.__session = response['user_token']
        with open(self.__session_file, 'w+') as f:
            f.write(self.__session)

    def _account(self):
        return self._refreshed(lambda: self._get_req('account', cookies=self._cookie_dict))

    @staticmethod
    def _id_from_url(id_or_url):
        """Handles grabbing the Eero ID from the URL"""
        match = re.search('^[0-9]+$', id_or_url)
        if match:
            return match.group(0)
        match = re.search(r'\/([0-9]+)$', id_or_url)
        if match:
            return match.group(1)

    def _devices(self, network_id):
        """Gets the list of devices from Eero"""
        return self._refreshed(lambda: self._get_req('networks/{}/devices'.format(self._id_from_url(network_id)),
                                                     cookies=self._cookie_dict))

    @staticmethod
    def _parse_response(response):
        """Basic response handler"""
        data = json.loads(response.text)
        if data['meta']['code'] is not 200 and data['meta']['code'] is not 201:
            raise EeroException(data['meta']['code'], data['meta'].get('error', ""))
        return data.get('data', "")

    def _post_req(self, action, **kwargs):
        """POST a request"""
        response = requests.post(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)

    def _get_req(self, action, **kwargs):
        """GET a request"""
        response = requests.get(self.API_ENDPOINT.format(action), **kwargs)
        return self._parse_response(response)
